django/utils/deprecation.py

```diff
@@ -83,6 +85,187 @@ def __new__(cls, name, bases, attrs):
+                " (Apply it to the __init__ method.)"
+            )
+        if isinstance(func, classmethod):
+            raise TypeError("Apply @classmethod before @deprecate_posargs.")
+        if isinstance(func, staticmethod):
+            raise TypeError("Apply @staticmethod before @deprecate_posargs.")
+
+        params = inspect.signature(func).parameters
+        num_by_kind = Counter(param.kind for param in params.values())
+
+        if num_by_kind[inspect.Parameter.VAR_POSITIONAL] > 0:
+            raise TypeError(
+                "@deprecate_posargs() cannot be used with variable positional `*args`."
+            )
+
+        num_positional_params = (
+            num_by_kind[inspect.Parameter.POSITIONAL_ONLY]
+            + num_by_kind[inspect.Parameter.POSITIONAL_OR_KEYWORD]
+        )
+        num_keyword_only_params = num_by_kind[inspect.Parameter.KEYWORD_ONLY]
+        if num_keyword_only_params < 1:
+            raise TypeError(
+                "@deprecate_posargs() requires at least one keyword-only parameter"
+                " (after a `*` entry in the parameters list)."
+            )
+        if any(
+            name not in params or params[name].kind != inspect.Parameter.KEYWORD_ONLY
+            for name in remappable_names
+        ):
+            raise TypeError(
+                "@deprecate_posargs() remappable_names must"
+                " all be keyword-only parameters."
+            )
+
+        num_remappable_args = len(remappable_names)
+        max_positional_args = num_positional_params + num_remappable_args
+
+        func_name = func.__name__
+        if func_name == "__init__":
+            # In the warning, show "ClassName(...)" rather than "__init__(...)".
+            # The class isn't defined yet, but its name is in __qualname__.
+            # Some examples: "ClassName.__init__", "Nested.ClassName.__init__",
+            # "MyTests.test_case.<locals>.ClassName.__init__".
+            local_name = func.__qualname__.rsplit("<locals>.", 1)[-1]
+            class_name = local_name.replace(".__init__", "")
+            func_name = class_name
+
+        def remap_deprecated_args(args, kwargs):
+            """
+            Move deprecated positional args to kwargs and issue a warning.
+            Returns updated (args, kwargs).
+            """
+            num_positional_args = len(args)
+            if num_positional_args > max_positional_args:
+                raise TypeError(
+                    f"{func_name}() takes"
+                    f" at most {max_positional_args} positional argument(s)"
+                    f" (including {num_remappable_args} deprecated)"
+                    f" but {num_positional_args} were given."
+                )
+
+            # Identify which of the potentially _remappable_ params
+            # are actually being _remapped_ in this particular call.
+            remapped_names = remappable_names[
+                : num_positional_args - num_positional_params
```

Comment:
I tried to do some "mutation testing" on the code to have a second level of checks about tests, and notice that if we remove the subtracting argument (i.e., leaving only `: num_positional_args`), no test fail. Could you add a case coverging for this?